Refactor tailscale nginx-auth module

Dan-Theriault opened the pull request:
## Description of changes

Tailscale's nginx-auth proxy, despite the name, actually works just fine with caddy and traefik (see tailscale/tailscale#4680). However, the current nixos module is closely coupled with nginx. This change breaks that coupling.

All options of the existing module except `virtualHosts` (which is nginx-specific) have been moved to a new module without any nginx coupling. The options remain exposed on the nginx-specific module as aliases.

closes #290353 

@phaer 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Commit:
nixos/tailscale-auth: init module

This additional module allows the tailscale auth proxy to be configured
independently of nginx. The tailscale auth proxy works with both caddy
and traefik. All prior nginx/tailscale-auth options are retained as
aliases.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3618

SuperSandro2000 commented on nixos/modules/services/web-servers/nginx/tailscale-auth.nix:
```diff
@@ -4,43 +4,20 @@ with lib;
 
 let
   cfg = config.services.nginx.tailscaleAuth;
+  cfgAuth = config.services.networking.tailscaleAuth;
```
```suggestion
  cfgAuth = config.services.tailscaleAuth;
```

SuperSandro2000 commented on nixos/modules/services/networking/tailscale-auth.nix:
```diff
@@ -0,0 +1,107 @@
+      description = lib.mdDoc "Group which runs tailscale-nginx-auth";
+    };
+
+    expectedTailnet = mkOption {
```
That path is not even used in this module. Does it really fit here?

Dan-Theriault replied:
My thinking here is that I would like to build other modules layered over this, similar to the nginx module, and it would be convenient to have a universal option for this. I do see it is a odd & can remove it to get this merged.

SuperSandro2000 commented on nixos/modules/services/networking/tailscale-auth.nix:
```diff
@@ -0,0 +1,107 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
We should avoid this here

Dan-Theriault replied:
I've replaced this with the `inherit (lib) ...` pattern.

phaer reviewed (commented):
Good news! Thanks for bringing this to attention.

Do you think it would make sense to wait for results of the upstream issue you linked? I.e. merging this before a rename means, that the program is confusingly still called tailscale-nginx-auth and that it's missing the '/auth-forward' endpoint, suggested in https://github.com/tailscale/tailscale/issues/4680.

Dan-Theriault commented:
> Good news! Thanks for bringing this to attention.
> 
> Do you think it would make sense to wait for results of the upstream issue you linked? I.e. merging this before a rename means, that the program is confusingly still called tailscale-nginx-auth and that it's missing the '/auth-forward' endpoint, suggested in [tailscale/tailscale#4680](https://github.com/tailscale/tailscale/issues/4680).

[Caddy](https://caddyserver.com/docs/caddyfile/directives/forward_auth#tailscale) officially documents support for the feature. Their suggested workaround for lacking '/auth-forward` is just three lines of caddy config. I don't think that should prevent this from going ahead. The confusing program name is really the worst bit.

The current upstream situation isn't ideal, but I think it's good enough. There have not been many signs of life on the upstream ticket in the past year, and it is unclear to me if/when improvements are coming.

Dan-Theriault commented:
I think I've addressed all feedback now.

SuperSandro2000 commented on nixos/modules/services/networking/tailscale-auth.nix:
```diff
@@ -0,0 +1,105 @@
+      requires = [ "tailscale-nginx-auth.socket" ];
+
+      serviceConfig = {
+        ExecStart = "${getExe cfg.package}";
```
```suggestion
        ExecStart = getExe cfg.package;
```

MinerSebas commented on nixos/modules/services/web-servers/nginx/tailscale-auth.nix:
```diff
@@ -52,66 +46,14 @@ in
 
   config = mkIf cfg.enable {
     services.tailscale.enable = true;
+    services.tailscaleAuth.enable = true;
```
```suggestion
    services.tailscaleAuth.enable = true;
```
Nit: Technically unnecessary as `tailscaleAuth` already enables tailscale.

Dan-Theriault replied:
Thanks!

Dan-Theriault commented:
Got my maintainers entry added in another PR, so I've just rebased this.

Dan-Theriault commented:
I've rebased again to resolve conflicts with #303841. Removed a remaining `mkPackageOptionMD` while I was at it. Does anyone have further feedback? Not sure what's needed now to get this merged.
